modules/gdscript/gdscript_parser.cpp

```diff
@@ -4224,17 +4308,72 @@ bool GDScriptParser::export_group_annotations(const AnnotationNode *p_annotation
 
 bool GDScriptParser::warning_annotations(const AnnotationNode *p_annotation, Node *p_target, ClassNode *p_class) {
 #ifdef DEBUG_ENABLED
+	if (is_ignoring_warnings) {
+		return true; // We already ignore all warnings, let's optimize it.
+	}
+
 	bool has_error = false;
 	for (const Variant &warning_name : p_annotation->resolved_arguments) {
-		GDScriptWarning::Code warning = GDScriptWarning::get_code_from_name(String(warning_name).to_upper());
-		if (warning == GDScriptWarning::WARNING_MAX) {
+		GDScriptWarning::Code warning_code = GDScriptWarning::get_code_from_name(String(warning_name).to_upper());
+		if (warning_code == GDScriptWarning::WARNING_MAX) {
 			push_error(vformat(R"(Invalid warning name: "%s".)", warning_name), p_annotation);
 			has_error = true;
 		} else {
-			p_target->ignored_warnings.push_back(warning);
+			int start_line = p_annotation->start_line;
+			int end_line = p_target->end_line;
+
+			switch (p_target->type) {
+#define SIMPLE_CASE(m_type, m_class, m_property)          \
+	case Node::m_type: {                                  \
+		m_class *node = static_cast<m_class *>(p_target); \
+		if (node->m_property != nullptr) {                \
+			end_line = node->m_property->end_line;        \
+		} else {                                          \
+			end_line = node->start_line;                  \
+		}                                                 \
+	} break;
+
+				// Can contain properties (set/get).
+				SIMPLE_CASE(VARIABLE, VariableNode, initializer)
+
+				// Contain bodies.
+				SIMPLE_CASE(FOR, ForNode, list)
+				SIMPLE_CASE(IF, IfNode, condition)
+				SIMPLE_CASE(MATCH, MatchNode, test)
+				SIMPLE_CASE(WHILE, WhileNode, condition)
+#undef SIMPLE_CASE
+
+				case Node::CLASS: {
+					end_line = p_target->start_line;
+					for (const AnnotationNode *annotation : p_target->annotations) {
+						start_line = MIN(start_line, annotation->start_line);
+						end_line = MAX(end_line, annotation->end_line);
+					}
+				} break;
+
+					/* clang-format off */
+				/*case Node::FUNCTION: {
+					FunctionNode *function = static_cast<FunctionNode *>(p_target);
+					end_line = function->start_line;
+					for (int i = 0; i < function->parameters.size(); i++) {
+						end_line = function->parameters[i]->end_line;
+						if (function->parameters[i]->initializer != nullptr) {
+							end_line = function->parameters[i]->initializer->end_line;
+						}
+					}
+				} break;*/
+					/* clang-format on */
```

Comment:
We have a controversial feature: `@warning_ignore` applied to a function affects its entire body, not just the header (unused parameters, warnings on their initializers). Like here:

https://github.com/godotengine/godot/blob/b4e2a24c1f62088b3f7ce0197afc90832fc25009/modules/gdscript/tests/scripts/analyzer/warnings/shadowning.gd#L9-L17

Reply:
Controversial? I think it makes sense to me and reduces repetition. If `@warning_ignore` could be applied to the class itself it would make even more sense.

Reply:
For me, this is a controversial feature, since you can accidentally disable a warning for an entire function, although you intended to do it only for the parameter initializer.

Also, what about the other cases? In my opinion, ignoring a warning on a variable should not ignore warnings in its getter/setter, ignoring a warning on `if` should not ignore all warnings of that type in the entire block.

* Regarding mass-ignoring, see #76020.

Reply:
Sorry, we do not see it the same way. I think warnings being ignored in the whole scope makes more sense in GDScript, as it's more accessible and easier to read. This is so long as it's documented and the behavior is consistent.

> > since you can accidentally disable a warning for an entire function, although you intended to do it only for the parameter initializer.

There's a clear distinction between the `unused_parameter` and `unused_property` warning and their purpose does not intertwine. I find this difficult to happen in practice.

> ignoring a warning on a variable should not ignore warnings in its getter/setter, ignoring a warning on if should not ignore all warnings of that type in the entire block.

Wen assigning a value a setter/getter function is called and the scope is shifted to the called function. So of course warnings would be issued only within that nested context.
```GDscript
var property: int:
    set(new_value):
         # Only warn about standalone expression here.
         @warning_ignore("standalone_expression")
         "Standalone expression"
         property = new_value

func _ready():
    # Only warn about narrowing conversion here.
    @warning_ignore("narrowing_conversion")
    property = 1.5
```
When an `if` condition passes or similar, a nested scope is created, I imagine. The behavior sounds consistent, to me.
```GDScript
# I know what I am doing here, but I may REALLY need that warning anywhere else.
@warning_ignore("narrowing_conversion")
@warning_ignore("integer_division")
func _ready():
    int_height = height_input.value / 2
    int_width = width_input.value / 2
    int_depth = depth_input.value / 2
```

I would be in favour of it over the [two new annotations](https://github.com/godotengine/godot/pull/76020) (does not mean they do not have merit).

Reply:
> There's a clear distinction between the `unused_parameter` and `unused_property` warning and their purpose does not intertwine. I find this difficult to happen in practice.

This is not true for other warning types:

https://github.com/godotengine/godot/assets/47700418/83aa875f-d6f2-49f5-86be-c781a0202e45

Reply:
I have marked this as resolved because I am not changing it in this PR. For further discussion, see #76020.